### PR TITLE
DEV-1492: preserve transform kwargs in filter normalization

### PR DESCRIPTION
**Pull request by ZmeiGorynych**
## Summary
- `parse_filter_expr`'s blanket `=`→`==` rewrite mangled keyword arguments inside transform / parametric-aggregation calls, so `ntile(<m>, n=4) <= 1`, `rank(<m>, partition_by=<col>) <= N`, and `revenue:percentile(p=0.5) > 100` were rejected at the binder ("Transform 'X' accepts exactly one positional argument"). The documented top-N-in-filter patterns were broken end-to-end whenever a kwarg was present.
- Replace the per-part `=`→`==` regex with a literal-aware, paren-kind-tracking single scan that preserves `=` as a kwarg iff the innermost open paren is a CALL paren whose callee is not in `SCALAR_FUNCTIONS`, and the `=` matches Python's `IDENT preceded by ( or ,` kwarg grammar. Lowercase keywords (`and`/`or`/`not`/`in`/`is`) do not classify a following `(` as a call paren.
- The `SCALAR_FUNCTIONS` narrowing mirrors the architecture's existing no-kwargs-on-scalars rule (`docs/architecture/parsing.md`), so `coalesce(status = 'paid', False)` still normalizes to a `ScalarCall` with a `Cmp` arg as before — confirmed by a regression test.

Linear: [DEV-1492](https://linear.app/motley-ai/issue/DEV-1492/parse-filter-expr-mangles-transform-kwargs-in-filters-ntilex-n4-1)

## Test plan
- [x] `tests/test_syntax.py::TestFilterOperatorNormalization` — un-xfail tracking test + 12 new parse-level cases: `rank(..., partition_by=region)`, list form `partition_by=[a,b]`, grouping-paren guard, mixed transform-kwarg + top-level equality, literal-awareness `status = 'a=b'`, scalar-comparison preserved (`coalesce`/`ifnull`), colon-agg kwarg `revenue:percentile(p=0.5) > 100`, spaced kwarg `n = 4`, `not(...)` grouping classification, nested scalar-in-transform stacking, string literal containing `)`/`,`/`=`.
- [x] `tests/test_sql_generator.py::TestRankFamilyTransforms` — 2 end-to-end emission tests pinning `NTILE(4)` and `RANK() OVER (PARTITION BY ...)` to the inner SELECT and the `<= 1` predicate to the `_filtered` outer wrapper.
- [x] Full non-integration suite: `3917 passed, 6 skipped, 41 xfailed, 0 failed` (3m22s).
- [x] `ruff check slayer/ tests/` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved filter-expression parsing: `=` is now reliably converted to `==` where appropriate while preserving function keyword arguments and respecting grouping vs. call parentheses; string literals and compound operators remain unchanged.

* **Tests**
  * Expanded regression and unit tests covering nested calls, kwarg preservation, grouping vs. call behavior, scalar-function arguments, and string-literal edge cases.

### CODE REVIEW

File: docs/architecture/errors-and-warnings.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+### `ValueError` multi-inheritance for back-compat
+
+`UnknownReferenceError`, `AmbiguousReferenceError`, `IllegalScopeReferenceError`,
+and `IllegalWindowInFilterError` multi-inherit `ValueError` (alongside
+`SlayerError`). This is deliberate: the cutover replaced legacy `ValueError`
+resolution paths with these typed errors, and many pre-existing call sites and
+tests catch `ValueError` (or use `pytest.raises(ValueError)`). Multi-inheriting
+keeps them working unchanged. `ColumnCycleError` (DEV-1410) does the same.
+
+`SlayerError` is the base for SLayer's intentional failure modes, so callers can
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply docs/architecture/errors-and-warnings.md:49]: @coderabbitai This thread cites `docs/architecture/errors-and-warnings.md` — a file that is not part of PR #158's diff against its corrected base (`egor/dev-1484-dev-1452-stage-c-migrate-pre-existing-tests-off-legacy`). The thread was created when PR #158 originally targeted `main` and the whole dev-1484 stack appeared in the diff; after the base retarget those files belong to the parent stack branch, not this PR. Leaving as out-of-scope here — the fix (Slayer vs SLayer casing) is welcome on the parent branch's PR.

File: slayer/engine/binding.py [hunk 2]
```diff
@@ -0,0 +1,1296 @@
...
+    """
+    if isinstance(parsed, Literal):
+        return normalize_scalar(parsed.value)
+    if (
+        isinstance(parsed, UnaryOp)
+        and parsed.op == "-"
+        and isinstance(parsed.operand, Literal)
+    ):
+        from decimal import Decimal
+
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/binding.py:1001]: @coderabbitai This thread cites `slayer/engine/binding.py:1001` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `binding.py` belongs to the parent branch, not this PR. Out-of-scope here — the import-placement fix is welcome on the parent branch's PR.

File: slayer/engine/planning.py [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+            type=type,
+            expression=expression if expression is not None else BoundExpr(value_key=key),
+            format=format,
+            description=description,
+        )
+        self._slots[sid] = slot
+        self._by_key[key] = sid
+        if public_name is not None:
+            self._declared_names[public_name] = sid
+        return sid
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/planning.py:250]: @coderabbitai This thread cites `slayer/engine/planning.py:250` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `planning.py` belongs to the parent branch, not this PR. Out-of-scope here — the complexity refactor is welcome on the parent branch's PR.

File: slayer/engine/planning.py [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+        return BetweenKey(column=new_col, low=new_low, high=new_high)
+    if isinstance(key, InKey):
+        # DEV-1475: ``InKey.values`` is a literal-only tuple, so it
+        # carries no sugar to lower; only the LHS column can host a
+        # rewritable transform. Rebuild only if the column changed.
+        new_col = lower_sugar_transforms(key.column)
+        if new_col is key.column:
+            return key
+        return InKey(column=new_col, values=key.values, negated=key.negated)
+    return key
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/planning.py:400]: @coderabbitai This thread cites `slayer/engine/planning.py:400` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `planning.py` belongs to the parent branch, not this PR. Out-of-scope here — the `lower_sugar_transforms` split is welcome on the parent branch's PR.

File: slayer/engine/planning.py [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+                        hidden=True,
+                        phase=dep.phase,
+                    )
+
+        return ProjectionPlan(
+            registry=registry,
+            public_projection=public_projection,
+            filters=filters,
+            order=order,
+        )
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/planning.py:627]: @coderabbitai This thread cites `slayer/engine/planning.py:627` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `planning.py` belongs to the parent branch, not this PR. Out-of-scope here — the `ProjectionPlanner.plan` deduplication is welcome on the parent branch's PR.

File: slayer/engine/stage_ordering.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    hidden inside a nested stage still surfaces as an edge from the
+    enclosing stage. Same recursion through ``ModelExtension`` shapes.
+
+    The traversal is purely structural — it never touches storage and
+    never raises on a missing-sibling reference (the caller validates
+    edges against ``against``).
+    """
+    out: Set[str] = set()
+    _walk_spec(query.source_model, against, out)
+    return out
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/stage_ordering.py:48]: @coderabbitai This thread cites `slayer/engine/stage_ordering.py:48` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `stage_ordering.py` belongs to the parent branch, not this PR. Out-of-scope here — the kwarg-call style fix is welcome on the parent branch's PR.

File: slayer/engine/stage_planner.py [hunk 7]
```diff
@@ -0,0 +1,1572 @@
...
+    if isinstance(key, InKey):
+        # DEV-1475: ``InKey.values`` is a literal-only tuple — no
+        # transforms to attach a time key to. Only the LHS column path
+        # can carry a transform; rebuild only if it changed.
+        nc = _attach_time_keys(key.column, td_key=td_key)
+        if nc is key.column:
+            return key
+        return InKey(column=nc, values=key.values, negated=key.negated)
+    return key
+
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/stage_planner.py:172]: @coderabbitai This thread cites `slayer/engine/stage_planner.py:172` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `stage_planner.py` belongs to the parent branch, not this PR. Out-of-scope here — the planner-helpers complexity refactor is welcome on the parent branch's PR.

File: slayer/memories/resolver.py [hunk 8]
```diff
@@ -528,28 +548,24 @@ def _add(forms: Iterable[str]) -> None:
...
-        if m.name is not None
-    }
-    extra_agg_names = frozenset(
-        a.name for a in source_model.aggregations
-    ) | BUILTIN_AGGREGATIONS
+    # 4. measures — parse each formula (Mode-B DSL) and resolve each
+    # referenced entity token. Function-style aggregations are rewritten to
+    # colon syntax first (quiet FUNC_STYLE_AGG slack helper), including the
+    # source model's custom aggregations — matching the legacy parse path.
+    custom_agg_names = frozenset(a.name for a in source_model.aggregations)
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/memories/resolver.py:555]: @coderabbitai This thread cites `slayer/memories/resolver.py:555` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `resolver.py` belongs to the parent branch, not this PR. Out-of-scope here — the defensive `or []` is welcome on the parent branch's PR.

File: tests/test_binding.py [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+    def test_flat_underscore_name_resolves(self):
+        # The flat name `robot_details__modelseriesval` IS a column on
+        # the stage schema — it resolves.
+        scope = _stage_schema_with_flat_names()
+        # The syntax parser rejects `__` in user identifiers — bind_expr
+        # is given a pre-parsed ParsedExpr that explicitly uses the flat
+        # column name via the Ref constructor.
+        from slayer.engine.syntax import Ref
+        bound = bind_expr(Ref(name="robot_details__modelseriesval"),
+                          scope=scope, bundle=_bundle())
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_binding.py:415]: @coderabbitai This thread cites `tests/test_binding.py:415` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `test_binding.py` belongs to the parent branch, not this PR. Out-of-scope here — the test-import placement fix is welcome on the parent branch's PR.

File: tests/test_boundexpr_unification.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        q = SlayerQuery(
+            source_model="orders",
+            measures=[{"formula": "amount:sum"}],
+            filters=["amount > 0"],
+        )
+        planned = plan_query(query=q, bundle=_bundle())
+        fp = planned.filters_by_phase[0]
+        # Should be an ArithmeticKey wrapping ColumnKey(amount) and a
+        # LiteralKey.
+        from slayer.core.keys import ArithmeticKey
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_boundexpr_unification.py:161]: @coderabbitai This thread cites `tests/test_boundexpr_unification.py:161` — not part of PR #158's diff against its corrected base (`egor/dev-1484-...`). The thread was created when PR #158 originally targeted `main` and the dev-1484 stack appeared in the diff; after the retarget, `test_boundexpr_unification.py` belongs to the parent branch, not this PR. Out-of-scope here — the test-import placement fix is welcome on the parent branch's PR.
